### Experiment
#### 1.不同线搜索准则的比较
##### 1.1 Watson

In [1]:
from module import *
import time
watson = Watson(31) 

In [2]:
print('### Goldstein ###')
begin = time.time()
DampedNewton([0.,0.],watson,Goldstein())
end = time.time()
print('Time Consumption: ',end - begin)

### Goldstein ###
迭代结束-iter: 5406  x: [-0.60101944  1.08934779]  loss: 0.33219668594973284
Time Consumption:  27.489601373672485


In [3]:
print('### Wolfe ###')
begin = time.time()
DampedNewton([0.,0.],watson,Wolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Wolfe ###
迭代结束-iter: 5800  x: [-0.60101944  1.08934779]  loss: 0.33219668594973256
Time Consumption:  36.97166919708252


In [4]:
print('### Strong Wolfe ###')
begin = time.time()
DampedNewton([0.,0.],watson,StrongWolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Strong Wolfe ###
迭代结束-iter: 5735  x: [-0.60101944  1.08934779]  loss: 0.33219668594973284
Time Consumption:  32.67377710342407


- 分析 <br/>
选择需要通过线搜索确定步长的阻尼Newton法，比较不同线搜索方法的差异。<br/>
在Watson函数上，Goldstein方法经过5406次迭代，收敛到最优解，消耗的CPU时间为27.4896s；Wolfe方法经过5800次迭代，收敛到最优解，消耗的CPU时间为36.97117s；Strong Wolfe方法经过5735次迭代，收敛到最优解，消耗的CPU时间为32.6738s。Goldstein所需的迭代次数和时间最少，在该实验中优于另外两种搜索方法。Wolfe和Strong Wolfe函数的迭代次数接近，但是Strong Wolfe所用的的时间更少。这是因为在代码的实现过程中，Strong Wolfe条件仅比Wolfe条件多一次求绝对值的运算，就能获得更加精确的搜索步长。

##### 1.2 Discrete Boundary Value (2 dimension)

In [5]:
dbv = DiscreteBoundaryValue(2)
print('### Goldstein ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Goldstein())
end = time.time()
print('Time Consumption: ',end - begin)

### Goldstein ###
迭代结束-iter: 18  x: [-6.07549555e+11 -1.15651758e+07]  loss: 1.552191252985048e+68
Time Consumption:  0.012641429901123047


In [6]:
print('### Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Wolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Wolfe ###
迭代结束-iter: 18  x: [-6.07549555e+11 -1.15651758e+07]  loss: 1.552191252985048e+68
Time Consumption:  0.012243986129760742


In [7]:
print('### Strong Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,StrongWolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Strong Wolfe ###
迭代结束-iter: 18  x: [-6.07549555e+11 -1.15651758e+07]  loss: 1.552191252985048e+68
Time Consumption:  0.014055013656616211


##### 1.3 Discrete Boundary Value (5 dimension)

In [8]:
dbv = DiscreteBoundaryValue(5)
print('### Goldstein ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Goldstein())
end = time.time()
print('Time Consumption: ',end - begin)

### Goldstein ###
迭代结束-iter: 9  x: [  20.33954345   -3.20149727 -422.53735813 -112.21796325  650.03711123]  loss: 15888779839877.893
Time Consumption:  0.014214038848876953


In [9]:
print('### Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Wolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Wolfe ###
迭代结束-iter: 9  x: [  20.33954345   -3.20149727 -422.53735813 -112.21796325  650.03711123]  loss: 15888779839877.893
Time Consumption:  0.014165401458740234


In [10]:
print('### Strong Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,StrongWolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Strong Wolfe ###
迭代结束-iter: 9  x: [  20.33954345   -3.20149727 -422.53735813 -112.21796325  650.03711123]  loss: 15888779839877.893
Time Consumption:  0.013999700546264648


##### 1.4 Discrete Boundary Value (10 dimension)

In [11]:
dbv = DiscreteBoundaryValue(10)
print('### Goldstein ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Goldstein())
end = time.time()
print('Time Consumption: ',end - begin)

### Goldstein ###
迭代结束-iter: 8  x: [   273.03577174   -318.79766238    698.75182756   -324.31613194
    176.201222     -399.46482776    333.71585835   -232.57754566
 -12091.69513813  -2186.50837798]  loss: 5.332345385085355e+19
Time Consumption:  0.02046060562133789


In [12]:
print('### Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,Wolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Wolfe ###
迭代结束-iter: 8  x: [   273.03577174   -318.79766238    698.75182756   -324.31613194
    176.201222     -399.46482776    333.71585835   -232.57754566
 -12091.69513813  -2186.50837798]  loss: 5.332345385085355e+19
Time Consumption:  0.015334606170654297


In [13]:
print('### Strong Wolfe ###')
begin = time.time()
DampedNewton(dbv.init_point(),dbv,StrongWolfe())
end = time.time()
print('Time Consumption: ',end - begin)

### Strong Wolfe ###
迭代结束-iter: 8  x: [   273.03577174   -318.79766238    698.75182756   -324.31613194
    176.201222     -399.46482776    333.71585835   -232.57754566
 -12091.69513813  -2186.50837798]  loss: 5.332345385085355e+19
Time Consumption:  0.020029067993164062


- 分析 <br>
选择需要通过线搜索确定步长的阻尼Newton法，比较不同线搜索方法的差异。<br/>
Discrete Boundary Value函数含有一个超参数$m$，表示损失函数$r_i(x)$的数量，在上述模拟实验中，分别取$m$的值为2，3，5进行实验。结果显示，在不同的参数设定下，Goldstein，Wolfe和Strong Wolfe搜索方法都因为出现Hessian矩阵不可逆的情况而无法求出最优解，仅经过有限次的迭代就终止了搜索过程，且搜索停止时的损失函数极大，这是因为在线搜索没有搜索到理想的步长时，会默认地采用一个较小的步长进行更新，导致迭代可能沿着梯度上升的方向更新。<br/>
同时随着参数$m$的增大，优化终止时的迭代次数减小。说明$m$的增加会使Discrete Boundary Value函数更容易出现矩阵奇异的情况。

##### 1.5结论

在数值模拟中，Goldstein、Wolfe和Strong Wolfe三种步长搜索方法在减少优化迭代次数、降低优化所需时间方面所起的左右较为接近；在代码实现上，Strong Wolfe能比Wolfe准则精准地搜索步长，因此使搜索算法更快地收敛；但是三种算法都无法解决Hessian矩阵不可逆的情况。

#### 不同搜索方法的比较

In [15]:
from module import *
import time
watson = Watson(31) 
dbv2 = DiscreteBoundaryValue(2)
dbv5 = DiscreteBoundaryValue(5)
dbv10 = DiscreteBoundaryValue(10)

##### 1.Newton法

In [9]:
print('### Watson ###')
begin = time.time()
BasicNewton([0.,0.],watson)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
BasicNewton(dbv2.init_point(),dbv2)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
BasicNewton(dbv5.init_point(),dbv5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
BasicNewton(dbv10.init_point(),dbv10)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 407  x: [-0.60101559  1.0893466 ]  loss: 0.33219668614111714
Time Consumption:  0.8028843402862549

### Discrete Boundary Value ###
	 iter: 3  x: [-6.20291648  7.44040953]  loss: 4664.522643235498
	 iter: 4  x: [-146.13161878 -336.11948191]  loss: 4350424187912.7007
迭代结束-iter: 5  x: [-1.21859101e+07 -3.47516632e+08]  loss: 5.43636417539642e+48
Time Consumption:  0.002142667770385742

### Discrete Boundary Value ###
	 iter: 3  x: [-2.32260613  2.10595278  0.012437   -2.88269929  3.13914358]  loss: 294.90204167395495
迭代结束-iter: 4  x: [-124.9148647   -29.29562094  -11.52446205 -107.55834866  -40.18913294]  loss: 982404486.6058675
Time Consumption:  0.0029985904693603516

### Discrete Boundary Value ###
	 iter: 3  x: [-1.89264449  1.56074825 -0.62870692 -0.22415186 -0.238988   -0.23699629
 -0.21829215  0.39659942 -2.3532075   2.56053027]  loss: 200.50504409783764
迭代结束-iter: 4  x: [-340.19899138  -50.96878041  104.47802861  -22.10140267    2.55670762
   -3.93334849

- 分析 <br>
分四种情形讨论基本牛顿法的收敛性质。 <br>
1.在Watson函数上，基本Newton法经过407次迭代，收敛到最优解，所用的CPU时间为0.8029s； <br>
2.对Discrete Boundary Value函数，无论$m=2, 3, 5$时，Hessian矩阵都多次出现非正定的情况，最终都因为Hessian矩阵奇异，不可求逆而终止搜索过程。迭代终止时，所用的迭代次数分别为5、4、4，消耗的时间分别为0.002143s,0.002998s,0.004544。可以看出，随着超参数$m$的增加，需要计算的损失函数数量和梯度、二次倒数的次数也增加，因此消耗的CPU时间也相应增加。

##### 2.SR1方法

In [16]:
print('### Watson ###')
begin = time.time()
SR1([0.,0.],watson)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
SR1(dbv2.init_point(),dbv2)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
SR1(dbv5.init_point(),dbv5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
SR1(dbv10.init_point(),dbv10)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 1000  x: [-0.76525263  1.2404707 ]  loss: 0.8860144037258358
Time Consumption:  3.144169569015503

### Discrete Boundary Value ###
迭代结束-iter: 655  x: [16920.292151    2782.56039842]  loss: 7.2462978293696435e+22
Time Consumption:  0.26785969734191895

### Discrete Boundary Value ###
迭代结束-iter: 6  x: [-2682044.01288278  2180271.80911787   131237.97208718 -2473178.05455994
  3133879.10176736]  loss: 3.1940314726501953e+35
Time Consumption:  0.006284236907958984

### Discrete Boundary Value ###
迭代结束-iter: 5  x: [-12934047.04589452  10314560.92586945  -2073906.3211809
   -276260.11741359     63020.53906735    -31474.56269032
    417719.0279904    2826205.65226695 -13426665.78608871
  16743602.51989637]  loss: 5.767938799420222e+38
Time Consumption:  0.010005474090576172


- 分析 <br>
分四种情形讨论SR1法的收敛性质。 <br>
1.在Watson函数上，基本Newton法经过1000次迭代，也没有收敛到最优解。消耗的CPU时间为3.1442s，迭代终止时损失函数的值为0.8860。
2.在Discrete Boundary Value函数上，$m=2, 5, 10$三种情形都因为损失函数值过大出现溢出而终止。这说明SR1方法作为一阶的Hessian举证更新方法，并不能够保证迭代过程中Hessian矩阵的正定性。这三种情形下的迭代次数依次为655, 6, 5；消耗的CPU时间依次为0.2678s, 0.006284s, 0.010005s。随着$m$的增加，discrete boundary value函数更容易出现损失函数值溢出的情况，因此$m=2$时函数终止所用的迭代次数和CPU时间最多。而在$m=5或10$时，由于后者需要计算的损失函数数量和梯度、二阶导数的次数更多，因此所用的时间更长。

##### 3.DFP方法 

In [14]:
print('### Watson ###')
begin = time.time()
DFP([0.,0.],watson)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
DFP(dbv2.init_point(),dbv2)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
DFP(dbv5.init_point(),dbv5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
DFP(dbv10.init_point(),dbv10)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 28  x: [-0.60101948  1.0893478 ]  loss: 0.3321966859497136
Time Consumption:  0.18523764610290527

### Discrete Boundary Value ###
迭代结束-iter: 8  x: [-0.12824676 -0.15926759]  loss: 3.694044736577643e-15
Time Consumption:  0.0049326419830322266

### Discrete Boundary Value ###
迭代结束-iter: 10  x: [-0.07503015 -0.13198611 -0.16485803 -0.16467292 -0.11742331]  loss: 7.331027265892723e-11
Time Consumption:  0.011516809463500977

### Discrete Boundary Value ###
	 iter: 9  x: [-0.06215008 -0.12012163 -0.17246142 -0.21067199 -0.23328838 -0.24205562
 -0.23629715 -0.21243424 -0.16692995 -0.09865159]  loss: 0.00037994452224853343
迭代结束-iter: 9  x: [-0.06215008 -0.12012163 -0.17246142 -0.21067199 -0.23328838 -0.24205562
 -0.23629715 -0.21243424 -0.16692995 -0.09865159]  loss: 0.00037994452224853343
Time Consumption:  0.017140865325927734


- 分析<br>
分四种情形讨论DFP法的收敛性质。 <br>
1.在Watson函数上，DFP方法经过28次迭代，收敛到最优解，所用的CPU时间为0.1852s; <br>
2.在Discrete Boundary Value函数上，$m=2$时，DFP方法经过8次迭代收敛到最优解，所用的时间为0.004933s；$m=5$时DFP方法经过10次迭代收敛到最优解，所用的时间为0.01152s；$m=10$时DFP因为碰到Hessian矩阵非奇异的情况，因此经过9次迭代后终止，所用的时间为0.01714s。<br>
DFP方法的收敛性质和收敛效率明显好于SR1方法，但是依然存在遇到奇异矩阵而无法收敛的问题。

##### 4.BFGS方法

In [15]:
print('### Watson ###')
begin = time.time()
BFGS([0.,0.],watson)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
BFGS(dbv2.init_point(),dbv2)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
BFGS(dbv5.init_point(),dbv5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
BFGS(dbv10.init_point(),dbv10)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 12  x: [-0.60102038  1.0893482 ]  loss: 0.3321966859602461
Time Consumption:  0.10546350479125977

### Discrete Boundary Value ###
迭代结束-iter: 5  x: [-0.12825474 -0.15927639]  loss: 2.5141912240305507e-10
Time Consumption:  0.0059108734130859375

### Discrete Boundary Value ###
迭代结束-iter: 12  x: [-0.07502282 -0.13198137 -0.1648639  -0.16468787 -0.11744043]  loss: 7.970154933735613e-10
Time Consumption:  0.014647722244262695

### Discrete Boundary Value ###
迭代结束-iter: 16  x: [-0.04316505 -0.08157733 -0.11448602 -0.14097398 -0.15990917 -0.16987767
 -0.16909036 -0.15524982 -0.12535605 -0.07541659]  loss: 2.443100744443053e-14
Time Consumption:  0.03131747245788574


- 分析<br>
分四种情形讨论BFGS法的收敛性质。 <br>
1.在Watson函数上，DFP方法经过12次迭代，收敛到最优解，所用的CPU时间为0.1054s，优于DFP方法；<br>
2.在Discrete Boundary Value函数上，$m=2$时，BFGS方法经过5次迭代收敛到最优解，所用的时间为0.00591s；$m=5$时BFGS方法经过12次迭代收敛到最优解，所用的时间为0.01464s；$m=10$时BFGS方法经过16次迭代收敛，所用的时间为0.03132s。可以看出随着$m$的增大，BFGS方法收敛所用的迭代次数和CPU时间都有所增加。<br>
相比之下，BFGS方法的表现效果要由于DFP方法，因为前者能保证Hessian矩阵可逆，使矩阵更可能收敛到函数的最优值；而且在算法都收敛的情况下，所用的CPU时间大致相同。

##### 5.Broyden族函数

In [16]:
## phi=0.5
print('### Watson ###')
begin = time.time()
Broyden([0.,0.],watson,0.5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
Broyden(dbv2.init_point(),dbv2,0.5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv5.init_point(),dbv5,0.5)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv10.init_point(),dbv10,0.5)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 16  x: [-259.65076274   38.42027132]  loss: 101578810601.45082
Time Consumption:  0.11752080917358398

### Discrete Boundary Value ###
迭代结束-iter: 51  x: [64.29928772 -7.62758371]  loss: 251008701.0968382
Time Consumption:  0.02646160125732422

### Discrete Boundary Value ###
迭代结束-iter: 20  x: [1.60687343 4.35466723 5.76036236 6.95681846 4.09284917]  loss: 231.54660194448942
Time Consumption:  0.018126249313354492

### Discrete Boundary Value ###
迭代结束-iter: 36  x: [ 18.24554722  25.32968835   6.94402696  -6.71116799 -36.26261295
 -40.55317937 -19.2532492   28.23269303  62.73701205  53.62718136]  loss: 2044067.3888920862
Time Consumption:  0.05665946006774902


In [17]:
## phi=0.1
print('### Watson ###')
begin = time.time()
Broyden([0.,0.],watson,0.1)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
Broyden(dbv2.init_point(),dbv2,0.1)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv5.init_point(),dbv5,0.1)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv10.init_point(),dbv10,0.1)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 86  x: [-93.08163314  34.41050451]  loss: 1038601986.4006084
Time Consumption:  0.3903796672821045

### Discrete Boundary Value ###
迭代结束-iter: 55  x: [1254.33865678  873.42081044]  loss: 1.348436061573599e+16
Time Consumption:  0.025806903839111328

### Discrete Boundary Value ###
迭代结束-iter: 42  x: [-187.07643939 -238.74591732 -323.6024834  -426.72725759 -346.89047068]  loss: 1721953585438.9329
Time Consumption:  0.04006338119506836

### Discrete Boundary Value ###
迭代结束-iter: 110  x: [ 8.05278424 21.00903983 25.1305864  38.35339331 50.4769221  49.39062336
 -0.52863609 -3.28297318  4.2043682  12.88469172]  loss: 782868.3339657588
Time Consumption:  0.16328907012939453


In [18]:
## phi=0.9
print('### Watson ###')
begin = time.time()
Broyden([0.,0.],watson,0.9)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
Broyden(dbv2.init_point(),dbv2,0.9)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv5.init_point(),dbv5,0.9)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv10.init_point(),dbv10,0.9)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 4  x: [-192.61212094  746.05585829]  loss: 695221269007.1653
Time Consumption:  0.03171658515930176

### Discrete Boundary Value ###
迭代结束-iter: 30  x: [ -8.15615378 -18.90711524]  loss: 99041.61723454553
Time Consumption:  0.023681640625

### Discrete Boundary Value ###
迭代结束-iter: 7  x: [-104661.78154525  -43171.81247471   -7685.30683773  178058.63041725
   39210.63973064]  loss: 6.40353569290962e+27
Time Consumption:  0.010741472244262695

### Discrete Boundary Value ###
迭代结束-iter: 7  x: [ 8324.11341248  3919.99477566 -1406.35753314    36.83660907
   181.89307968   190.02227782   382.19208267  2270.33531789
  1177.93952441 -1607.1664745 ]  loss: 5.750055874737652e+18
Time Consumption:  0.016606807708740234


In [19]:
## phi=0.0
print('### Watson ###')
begin = time.time()
Broyden([0.,0.],watson,0.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
Broyden(dbv2.init_point(),dbv2,0.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv5.init_point(),dbv5,0.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv10.init_point(),dbv10,0.)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 23  x: [-0.60101764  1.08934762]  loss: 0.33219668599682184
Time Consumption:  0.1679084300994873

### Discrete Boundary Value ###
迭代结束-iter: 7  x: [-0.12823791 -0.15926243]  loss: 2.2798490551110224e-10
Time Consumption:  0.005368471145629883

### Discrete Boundary Value ###
迭代结束-iter: 11  x: [-0.07502795 -0.13198408 -0.16485667 -0.1646716  -0.11742215]  loss: 3.650640137356127e-11
Time Consumption:  0.011769533157348633

### Discrete Boundary Value ###
迭代结束-iter: 16  x: [-0.04308268 -0.08141647 -0.11425574 -0.14069061 -0.15959644 -0.16956408
 -0.16880491 -0.15501731 -0.12519377 -0.07533374]  loss: 5.851113533280999e-09
Time Consumption:  0.032514333724975586


In [20]:
## phi=1.0
print('### Watson ###')
begin = time.time()
Broyden([0.,0.],watson,1.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

print('### Discrete Boundary Value ###')
begin = time.time()
Broyden(dbv2.init_point(),dbv2,1.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv5.init_point(),dbv5,1.)
end = time.time()
print('Time Consumption: ',end - begin)
print()

begin = time.time()
print('### Discrete Boundary Value ###')
Broyden(dbv10.init_point(),dbv10,1.)
end = time.time()
print('Time Consumption: ',end - begin)

### Watson ###
迭代结束-iter: 2  x: [-0.05491552  0.63792667]  loss: 7.8739368366206435
Time Consumption:  0.013666868209838867

### Discrete Boundary Value ###
迭代结束-iter: 8  x: [1481.73251216 -220.86020593]  loss: 3.2842590750799864e+16
Time Consumption:  0.0099029541015625

### Discrete Boundary Value ###
迭代结束-iter: 5  x: [ 340.67120039  694.46611472  255.26488409 -376.2811801  -155.57227106]  loss: 22797176605673.746
Time Consumption:  0.008309125900268555

### Discrete Boundary Value ###
迭代结束-iter: 5  x: [ 160.72603279  497.21879632  -91.04668164   35.53651426   43.89322951
   53.60106357   64.63295191  235.75485997 -530.05211718 -144.02844344]  loss: 638652337233.7291
Time Consumption:  0.013636589050292969


- 分析<br>
对Broyden函数族，对$\phi$取不同值的情况下对算法的收敛性进行讨论。<br>
1.在$\phi$依次取0.1, 0.5, 0.9和1.0时，算法都因为Hessian矩阵奇异而提前终止，没有收敛到最优值；<br>
2.而当$\phi=0.0$时，实际上执行的是BFGS方法。四种情形下算法都收敛。在Watson函数上，broyden方法经过23次迭代，消耗0.1679s的CPU时间收敛，在discrete boundary value($m=2$)函数上，方法经过7次迭代，消耗0.005368s的CPU时间收敛；在discrete boundary value($m=5$)函数上，方法经过11次迭代，消耗0.01176s的CPU时间收敛；在discrete boundary value($m=10$)函数上，方法经过16次迭代，消耗0.03251s的CPU时间收敛。

- 总结<br>
1.Newton方法由于需要计算Hessian矩阵的逆，因此相比于拟牛顿方法需要消耗的CPU时间更多，而且对Hessian奇异的情况无能为力；<br>
2.SR1方法，作为1阶的逆牛顿方法，有可能出现Hessian矩阵正定或奇异的情况，到时算法无法向损失函数下降的方向进行更行，甚至不得不提前终止。<br>
3.DFP和BFGD方法作为2阶逆牛顿方法，在求解效率上明显好于SR1方法。但是相比之下，BFGS算法由于存在自校正性质，故表现性能会好于DFP方法。
4.broyden函数族在参数$\phi=0.$是实际上退化为BFGS方法,在$\phi=1.0$是退化为DFP方法。